File: packages/server/dbos_argus/realtime/protocol.py
"""Wire protocol for the /ws endpoint.

One socket multiplexes many subscriptions, each tagged by a client-assigned
`sub_id`. Channel-specific payloads are opaque JSON and live in `data`.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

# Client → server


class SubscribeMessage(BaseModel):
    type: Literal["subscribe"] = "subscribe"
    sub_id: str
    channel: str
    params: dict[str, Any] | None = None


class UnsubscribeMessage(BaseModel):
    type: Literal["unsubscribe"] = "unsubscribe"
    sub_id: str


class UpdateParamsMessage(BaseModel):
    """Re-key an existing subscription without re-rendering on the client.

    Server may swap the underlying poller (in `KeyedChannel`) but the
    `sub_id` is preserved. Client gets a fresh `snapshot` for the new params.
    """

    type: Literal["update_params"] = "update_params"
    sub_id: str
    params: dict[str, Any] | None = None


class PingMessage(BaseModel):
    type: Literal["ping"] = "ping"


ClientMessage = SubscribeMessage | UnsubscribeMessage | UpdateParamsMessage | PingMessage


# Server → client


class SnapshotMessage(BaseModel):
    """First payload after `subscribe` (or after `update_params`)."""

    type: Literal["snapshot"] = "snapshot"
    sub_id: str
    channel: str
    data: Any
    # Server wall clock (epoch ms) at enqueue time. Lets clients compute a
    # clock-skew offset so UI geometry anchored on "now" (e.g. the timeline's
    # live edge) lines up with server-recorded timestamps.
    server_time_ms: int | None = None


class UpdateMessage(BaseModel):
    """Subsequent payload — sent only when the channel cursor advanced."""

    type: Literal["update"] = "update"
    sub_id: str
    channel: str
    data: Any
    server_time_ms: int | None = None


class ErrorMessage(BaseModel):
    type: Literal["error"] = "error"
    sub_id: str | None = None
    code: str
    message: str


class PongMessage(BaseModel):
    # Heartbeat pongs also carry the server clock so the skew offset stays
    # fresh on quiet connections with no data flowing.
    type: Literal["pong"] = "pong"
    server_time_ms: int | None = None


class AckMessage(BaseModel):
    """Confirms that a non-data control message landed (subscribe / unsubscribe).

    Useful for tests and for clients that want to know when an unsubscribe
    has been honored on the server.
    """

    type: Literal["ack"] = "ack"
    sub_id: str
    op: Literal["subscribe", "unsubscribe", "update_params"]


ServerMessage = SnapshotMessage | UpdateMessage | ErrorMessage | PongMessage | AckMessage


# Convenience top-level discriminated unions for parsing.


class IncomingEnvelope(BaseModel):
    """Discriminator-only model for parsing arbitrary client messages.

    Pydantic doesn't dispatch a tagged union from a top-level `BaseModel`
    parse call without a wrapper. We use this purely to validate the
    `type` discriminator before routing — actual payload parsing happens
    against the concrete model in the union.
    """

    type: str
    payload: dict[str, Any] = Field(default_factory=dict)

File: apps/console/src/lib/realtime/protocol.ts
/**
 * Wire protocol for the /ws endpoint.
 *
 * Mirrors `packages/server/dbos_argus/realtime/protocol.py`. Channel-specific
 * payload shapes are deliberately `unknown` here — typed wrappers in
 * `subscribe.svelte.ts` and channel-specific stores narrow them down.
 */

export type SubscribeMessage = {
  type: "subscribe";
  sub_id: string;
  channel: string;
  params?: Record<string, unknown>;
};

export type UnsubscribeMessage = {
  type: "unsubscribe";
  sub_id: string;
};

export type UpdateParamsMessage = {
  type: "update_params";
  sub_id: string;
  params?: Record<string, unknown>;
};

export type PingMessage = { type: "ping" };

export type ClientMessage =
  | SubscribeMessage
  | UnsubscribeMessage
  | UpdateParamsMessage
  | PingMessage;

export type SnapshotMessage = {
  type: "snapshot";
  sub_id: string;
  channel: string;
  data: unknown;
};

export type UpdateMessage = {
  type: "update";
  sub_id: string;
  channel: string;
  data: unknown;
};

export type ErrorMessage = {
  type: "error";
  sub_id?: string | null;
  code: string;
  message: string;
};

export type PongMessage = { type: "pong" };

export type AckMessage = {
  type: "ack";
  sub_id: string;
  op: "subscribe" | "unsubscribe" | "update_params";
};

export type ServerMessage =
  | SnapshotMessage
  | UpdateMessage
  | ErrorMessage
  | PongMessage
  | AckMessage;

export type ConnectionStatus = "closed" | "connecting" | "open";
